File: internal/format/format.go
package format

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/fatih/color"
	"github.com/nijaru/tk-go/internal/priority"
	"github.com/nijaru/tk-go/internal/task"
	"github.com/nijaru/tk-go/internal/timeutil"
)

// Colors
var (
	Green  = color.New(color.FgGreen).SprintFunc()
	Red    = color.New(color.FgRed).SprintFunc()
	Yellow = color.New(color.FgYellow).SprintFunc()
	Dim    = color.New(color.Faint).SprintFunc()
	Bold   = color.New(color.Bold).SprintFunc()

	StatusColors = map[task.Status]*color.Color{
		task.StatusOpen:   color.New(color.FgBlue),
		task.StatusActive: color.New(color.FgCyan),
		task.StatusDone:   color.New(color.Faint),
	}

	PriorityColors = map[task.Priority]*color.Color{
		task.PriorityUrgent: color.New(color.FgRed, color.Bold),
		task.PriorityHigh:   color.New(color.FgRed),
		task.PriorityMedium: color.New(color.FgYellow),
		task.PriorityLow:    color.New(color.FgBlue),
		task.PriorityNone:   color.New(color.Faint),
	}

	OverdueColor = color.New(color.FgRed, color.Bold)
	DueSoonColor = color.New(color.FgYellow)
)

func ShouldUseColor() bool {
	if os.Getenv("NO_COLOR") != "" {
		return false
	}
	// fatih/color handles TTY detection automatically, but we can override if needed
	return !color.NoColor
}

func formatId(id string, maxLen int) string {
	dash := strings.LastIndex(id, "-")
	if dash == -1 {
		return truncate(id, maxLen)
	}
	project := id[:dash]
	ref := id[dash+1:]
	maxProjectLen := maxLen - len(ref) - 1
	if len(project) > maxProjectLen {
		project = truncate(project, maxProjectLen)
	}
	return fmt.Sprintf("%s-%s", project, ref)
}

func truncate(text string, maxLen int) string {
	if len(text) <= maxLen {
		return text
	}
	if maxLen <= 1 {
		return "…"
	}
	return text[:maxLen-1] + "…"
}

func FormatTaskRow(t *task.TaskWithMeta) string {
	id := fmt.Sprintf("%-11s", formatId(t.ID, 11))
	prio := fmt.Sprintf("%-4s", priority.Format(t.Priority))
	
	isOverdue := t.IsOverdue
	dueSoon := !isOverdue && t.DaysUntilDue != nil && *t.DaysUntilDue <= timeutil.DueSoonThreshold

	statusText := string(t.Status)
	if t.Status == task.StatusDone && t.CompletedAt != nil {
		statusText = fmt.Sprintf("done %s", timeutil.FormatRelative(*t.CompletedAt))
	}
	status := fmt.Sprintf("%-12s", statusText)
	title := truncate(t.Title, 50)

	if ShouldUseColor() {
		pc := PriorityColors[t.Priority]
		sc := StatusColors[t.Status]
		if isOverdue {
			sc = OverdueColor
		} else if dueSoon {
			sc = DueSoonColor
		}
		
		tc := color.New()
		if t.Status == task.StatusDone {
			tc = color.New(color.Faint)
		}

		return fmt.Sprintf("%s | %s | %s | %s",
			id,
			pc.Sprint(prio),
			sc.Sprint(status),
			tc.Sprint(title),
		)
	}

	markers := ""
	if isOverdue {
		markers += " [OVERDUE]"
	} else if dueSoon {
		if *t.DaysUntilDue == 0 {
			markers += " [due today]"
		} else {
			markers += fmt.Sprintf(" [due %dd]", *t.DaysUntilDue)
		}
	}

	return fmt.Sprintf("%s | %s | %s | %s%s", id, prio, status, title, markers)
}

func FormatTaskList(tasks []*task.TaskWithMeta, emptyHint string) string {
	if len(tasks) == 0 {
		if emptyHint == "" {
			return "No tasks found. Run 'tk add \"title\"' to create one."
		}
		return emptyHint
	}

	header := "ID          | PRIO | STATUS       | TITLE"
	divider := strings.Repeat("-", 65)
	
	var rows []string
	rows = append(rows, header, divider)
	for _, t := range tasks {
		rows = append(rows, FormatTaskRow(t))
	}

	return strings.Join(rows, "\n")
}

func FormatTaskDetail(t *task.TaskWithMeta) string {
	var lines lineList

	sc := color.New()
	if c, ok := StatusColors[t.Status]; ok {
		sc = c
	}
	if t.IsOverdue {
		sc = OverdueColor
	}

	pc := color.New()
	if c, ok := PriorityColors[t.Priority]; ok {
		pc = c
	}

	lines = append(lines, fmt.Sprintf("ID:          %s", t.ID))
	lines.pushIfNonEmpty("Title:       ", t.Title)
	lines = append(lines, fmt.Sprintf("Status:      %s", sc.Sprint(string(t.Status))))
	lines = append(lines, fmt.Sprintf("Priority:    %s", pc.Sprint(priority.FormatName(t.Priority))))

	if t.Description != nil {
		lines = append(lines, fmt.Sprintf("Description: %s", *t.Description))
	}

	if len(t.Labels) > 0 {
		lines = append(lines, fmt.Sprintf("Labels:      %s", strings.Join(t.Labels, ", ")))
	}

	if len(t.Assignees) > 0 {
		lines = append(lines, fmt.Sprintf("Assignees:   %s", strings.Join(t.Assignees, ", ")))
	}

	if t.Parent != nil {
		lines = append(lines, fmt.Sprintf("Parent:      %s", *t.Parent))
	}

	if t.Estimate != nil {
		lines = append(lines, fmt.Sprintf("Estimate:    %d", *t.Estimate))
	}

	if t.DueDate != nil {
		dueStr := *t.DueDate
		if t.IsOverdue {
			dueStr += OverdueColor.Sprint(" [OVERDUE]")
		} else if t.DaysUntilDue != nil && *t.DaysUntilDue <= timeutil.DueSoonThreshold {
			if *t.DaysUntilDue == 0 {
				dueStr += DueSoonColor.Sprint(" [due today]")
			} else {
				dueStr += DueSoonColor.Sprint(fmt.Sprintf(" [due %dd]", *t.DaysUntilDue))
			}
		}
		lines = append(lines, fmt.Sprintf("Due:         %s", dueStr))
	}

	lines = append(lines, fmt.Sprintf("Created:     %s", timeutil.FormatDate(t.CreatedAt)))
	lines = append(lines, fmt.Sprintf("Updated:     %s", timeutil.FormatDate(t.UpdatedAt)))

	if t.CompletedAt != nil {
		lines = append(lines, fmt.Sprintf("Completed:   %s", timeutil.FormatDate(*t.CompletedAt)))
	}

	if len(t.BlockedBy) > 0 {
		status := " (resolved)"
		if t.BlockedByIncomplete {
			status = " (blocked)"
		}
		lines = append(lines, fmt.Sprintf("Blockers:    %s%s", strings.Join(t.BlockedBy, ", "), status))
	}

	if len(t.Logs) > 0 {
		lines = append(lines, "", "Log:")
		for _, log := range t.Logs {
			lines = append(lines, fmt.Sprintf("  [%s] %s", timeutil.FormatDate(log.Ts), log.Msg))
		}
	}

	return strings.Join(lines, "\n")
}

// Helper for detail formatting
type lineList []string

func (l *lineList) pushIfNonEmpty(prefix, value string) {
	if value != "" {
		*l = append(*l, prefix+value)
	}
}

func FormatJson(data interface{}) string {
	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "{}"
	}
	return string(b)
}

func FormatConfig(config task.Config) string {
	var lines []string
	lines = append(lines, fmt.Sprintf("Version:     %d", config.Version))
	lines = append(lines, fmt.Sprintf("Project:     %s", config.Project))
	
	if config.CleanAfter.Enabled {
		lines = append(lines, fmt.Sprintf("Clean After: %d days", config.CleanAfter.Days))
	} else {
		lines = append(lines, "Clean After: disabled")
	}

	if len(config.Defaults.Labels) > 0 {
		lines = append(lines, fmt.Sprintf("Def Labels:  %s", strings.Join(config.Defaults.Labels, ", ")))
	}
	if len(config.Defaults.Assignees) > 0 {
		lines = append(lines, fmt.Sprintf("Def Assigns: %s", strings.Join(config.Defaults.Assignees, ", ")))
	}
	lines = append(lines, fmt.Sprintf("Def Prio:    %s", priority.FormatName(config.Defaults.Priority)))

	if len(config.Aliases) > 0 {
		lines = append(lines, "", "Aliases:")
		for k, v := range config.Aliases {
			lines = append(lines, fmt.Sprintf("  %-10s -> %s", k, v))
		}
	}

	return strings.Join(lines, "\n")
}

File: src/format.rs
//! Output formatting: task tables, detail views, config display, JSON.
//!
//! The table shows the 4-character alias — the handle people type — while the
//! detail view shows both the alias and the full ULID. JSON always carries both.

use std::io::IsTerminal as _;

use owo_colors::OwoColorize;

use crate::model::{Config, Priority, Status, TaskState, TaskView};
use crate::timeutil;

/// Color when stdout is a TTY and `NO_COLOR` is unset.
pub fn use_color() -> bool {
    std::env::var_os("NO_COLOR").is_none() && std::io::stdout().is_terminal()
}

/// Truncate to `max` characters (Unicode-safe), appending `…` when cut.
pub fn truncate(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        return s.to_owned();
    }
    if max <= 1 {
        return "…".to_owned();
    }
    let mut out: String = s.chars().take(max - 1).collect();
    out.push('…');
    out
}

fn paint(color: bool, text: &str, style: Style) -> String {
    if !color {
        return text.to_owned();
    }
    match style {
        Style::Plain => text.to_owned(),
        Style::Dim => text.dimmed().to_string(),
        Style::Red => text.red().to_string(),
        Style::RedBold => text.red().bold().to_string(),
        Style::Yellow => text.yellow().to_string(),
        Style::Blue => text.blue().to_string(),
        Style::Cyan => text.cyan().to_string(),
    }
}

#[derive(Clone, Copy)]
enum Style {
    Plain,
    Dim,
    Red,
    RedBold,
    Yellow,
    Blue,
    Cyan,
}

fn status_style(s: Status) -> Style {
    match s {
        Status::Open => Style::Blue,
        Status::Active => Style::Cyan,
        Status::Done => Style::Dim,
        Status::Deferred => Style::Plain,
        Status::Closed => Style::Plain,
    }
}

fn priority_style(p: Priority) -> Style {
    match p {
        Priority::Urgent => Style::RedBold,
        Priority::High => Style::Red,
        Priority::Medium => Style::Yellow,
        Priority::Low => Style::Blue,
        Priority::None => Style::Dim,
    }
}

/// Width of the handle column: the longest alias in the result set.
fn alias_width(tasks: &[TaskView]) -> usize {
    tasks
        .iter()
        .map(|t| t.task.alias.chars().count())
        .max()
        .unwrap_or(4)
        .clamp(4, 12)
}

fn format_task_row_w(t: &TaskView, color: bool, w: usize) -> String {
    let prio = format!("{:<4}", t.task.priority.short());
    let mut status_text = t.task.status.to_string();
    if t.task.status.is_terminal()
        && let Some(c) = &t.task.completed_at
    {
        status_text = format!("{} {}", t.task.status, timeutil::format_relative(c));
    }
    let status = format!("{status_text:<12}");
    let title = truncate(&t.task.title, 50);

    if color {
        let tc = if t.task.status == Status::Done {
            Style::Dim
        } else {
            Style::Plain
        };
        return format!(
            "{:<w$} | {} | {} | {}",
            t.task.alias,
            paint(color, &prio, priority_style(t.task.priority)),
            paint(color, &status, status_style(t.task.status)),
            paint(color, &title, tc),
        );
    }

    let mut markers = String::new();
    if t.task.is_archived() {
        markers += " [archived]";
    }
    if t.blocked_by_incomplete {
        markers += " [blocked]";
    }
    format!(
        "{:<w$} | {prio} | {status} | {title}{markers}",
        t.task.alias,
        w = w
    )
}

pub fn format_task_list(tasks: &[TaskView], empty_hint: &str, color: bool) -> String {
    if tasks.is_empty() {
        if empty_hint.is_empty() {
            return "No tasks found. Run 'tk add \"title\"' to create one.".to_owned();
        }
        return empty_hint.to_owned();
    }
    let w = alias_width(tasks);
    let header = format!("{:<w$} | PRIO | STATUS       | TITLE", "REF");
    let divider = "-".repeat(header.chars().count());
    let mut rows = vec![header, divider];
    rows.extend(tasks.iter().map(|t| format_task_row_w(t, color, w)));
    rows.join("\n")
}

pub fn format_task_detail(t: &TaskView, color: bool) -> String {
    let task: &TaskState = &t.task;
    let mut lines = Vec::new();
    lines.push(format!("ID:          {}", task.id));
    lines.push(format!("Ref:         {}", task.alias));
    if !task.legacy_aliases.is_empty() {
        lines.push(format!("Also known:  {}", task.legacy_aliases.join(", ")));
    }
    lines.push(format!("Project:     {}", task.project));
    if !task.title.is_empty() {
        lines.push(format!("Title:       {}", task.title));
    }
    lines.push(format!(
        "Status:      {}",
        paint(color, &task.status.to_string(), status_style(task.status))
    ));
    lines.push(format!(
        "Priority:    {}",
        paint(color, task.priority.name(), priority_style(task.priority))
    ));
    if let Some(d) = &task.description {
        lines.push(format!("Description: {d}"));
    }
    if !task.labels.is_empty() {
        lines.push(format!("Labels:      {}", task.labels.join(", ")));
    }
    if let Some(p) = &t.parent_ref {
        lines.push(format!("Parent:      {p}"));
    }
    lines.push(format!(
        "Created:     {}",
        timeutil::format_date(&task.created_at)
    ));
    lines.push(format!(
        "Updated:     {}",
        timeutil::format_date(&task.updated_at)
    ));
    if let Some(c) = &task.completed_at {
        lines.push(format!("Completed:   {}", timeutil::format_date(c)));
    }
    if let Some(a) = &task.archived_at {
        lines.push(format!("Archived:    {}", timeutil::format_date(a)));
    }
    lines.push(format!("Revision:    {}", t.rev));
    if !t.unresolved_blockers.is_empty() {
        lines.push(format!(
            "Unresolved:  {} (task missing from the store)",
            t.unresolved_blockers
                .iter()
                .map(|id| id.chars().take(8).collect::<String>())
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }
    if !task.blocked_by.is_empty() {
        let state = if t.blocked_by_incomplete {
            " (blocked)"
        } else {
            " (resolved)"
        };
        lines.push(format!("Blockers:    {}{state}", t.blocker_refs.join(", ")));
    }
    if let Some(c) = &task.checkpoint {
        lines.push(String::new());
        lines.push(format!("Checkpoint:  {c}"));
    }
    if !task.links.is_empty() {
        lines.push(format!("Links:       {}", task.links.join(", ")));
    }
    if !task.acceptance.is_empty() {
        lines.push(String::new());
        lines.push("Acceptance:".to_owned());
        for a in &task.acceptance {
            lines.push(format!("  - {a}"));
        }
    }
    if !task.evidence.is_empty() {
        lines.push(String::new());
        lines.push("Evidence:".to_owned());
        for e in &task.evidence {
            lines.push(format!("  - {e}"));
        }
    }
    if !task.logs.is_empty() {
        lines.push(String::new());
        lines.push("Log:".to_owned());
        for log in &task.logs {
            lines.push(format!(
                "  [{}] {}",
                timeutil::format_date(&log.ts),
                log.msg
            ));
        }
    }
    lines.join("\n")
}

/// Single-line warning (yellow when color is on).
pub fn warning(text: &str, color: bool) -> String {
    paint(color, text, Style::Yellow)
}

pub fn format_json<T: serde::Serialize>(value: &T) -> String {
    serde_json::to_string_pretty(value).unwrap_or_else(|_| "{}".to_owned())
}

pub fn format_config(config: &Config) -> String {
    let mut lines = vec![
        format!("Format:      {}", config.format),
        format!("Project:     {}", config.project),
    ];
    if config.clean_after.enabled {
        lines.push(format!("Clean After: {} days", config.clean_after.days));
    } else {
        lines.push("Clean After: disabled".to_owned());
    }
    if !config.defaults.labels.is_empty() {
        lines.push(format!(
            "Def Labels:  {}",
            config.defaults.labels.join(", ")
        ));
    }
    lines.push(format!("Def Prio:    {}", config.defaults.priority.name()));
    if let Some(aliases) = &config.aliases
        && !aliases.is_empty()
    {
        lines.push(String::new());
        lines.push("Aliases:".to_owned());
        for (name, path) in aliases {
            lines.push(format!("  {name:<10} -> {path}"));
        }
    }
    lines.push(String::new());
    lines.push(
        "Change a setting with: tk config set <project|priority|labels|clean-after> <value>"
            .to_owned(),
    );
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{LogEntry, TaskState};

    fn sample() -> TaskView {
        TaskView {
            task: TaskState {
                id: "01j8x0m5r7000000000000000a".into(),
                alias: "a7b3".into(),
                legacy_aliases: Vec::new(),
                project: "demo".into(),
                title: "Implement auth".into(),
                description: None,
                status: Status::Open,
                priority: Priority::Urgent,
                labels: vec![],
                parent: None,
                blocked_by: vec![],
                logs: vec![LogEntry {
                    ts: "2026-01-10T12:00:00.000000000Z".into(),
                    msg: "note".into(),
                }],
                checkpoint: None,
                links: Vec::new(),
                acceptance: Vec::new(),
                evidence: Vec::new(),
                created_at: "2026-01-10T12:00:00.000000000Z".into(),
                updated_at: "2026-01-10T12:00:00.000000000Z".into(),
                completed_at: None,
                archived_at: None,
            },
            rev: "1-aaaa:1:deadbeef".into(),
            blocked_by_incomplete: false,
            unresolved_blockers: Vec::new(),
            blocker_refs: Vec::new(),
            parent_ref: None,
        }
    }

    #[test]
    fn truncate_is_unicode_safe() {
        assert_eq!(truncate("héllo→world", 6), "héllo…");
        assert_eq!(truncate("abc", 5), "abc");
        assert_eq!(truncate("abcdef", 1), "…");
    }

    #[test]
    fn the_table_uses_the_alias_not_the_ulid() {
        let t = sample();
        let table = format_task_list(std::slice::from_ref(&t), "", false);
        assert!(table.contains("a7b3"), "{table}");
        assert!(
            !table.contains("01j8x0m5r7000000000000000a"),
            "the 26-character ID must not widen every row: {table}"
        );
    }

    #[test]
    fn a_blocked_task_is_marked() {
        let mut t = sample();
        t.blocked_by_incomplete = true;
        t.task.blocked_by = vec!["01m25qbfpr5ekbr9zxh0xc93kx".into()];
        t.blocker_refs = vec!["vp80".into()];
        let table = format_task_list(std::slice::from_ref(&t), "", false);
        assert!(table.contains("[blocked]"), "{table}");

        let detail = format_task_detail(&t, false);
        assert!(detail.contains("Blockers:    vp80 (blocked)"), "{detail}");
        assert!(
            !detail.contains("Blockers:    01m25qbf"),
            "the blocker line must not lead with a ULID: {detail}"
        );
    }

    #[test]
    fn detail_shows_both_handles_and_renders_timestamps() {
        let t = sample();
        let detail = format_task_detail(&t, false);
        assert!(detail.contains("01j8x0m5r7000000000000000a"), "{detail}");
        assert!(detail.contains("Ref:         a7b3"), "{detail}");
        // Nano timestamps must render, not pass through raw.
        assert!(
            !detail.contains("2026-01-10T12:00:00.000000000Z"),
            "{detail}"
        );
    }
}
